### ISSUE TITLE
transaction show command with TXID

### ISSUE THREAD

**Issue by buhtignew** [slimcoin-project/pacli#2]
This is follow up for the https://www.reddit.com/r/slimcoin/comments/18asmhk/comment/ki0vx62/?utm_source=share&utm_medium=web2x&context=3.

I've tested `transaction show` command with `--structure` flag and the output makes sense to me.

The  same command without any flag (with the TXID only) has the following output:

_No transaction with this label or TXID was stored._

- which sounds a bit weird to me since I've just got the correct output using the `--structure` flag for the same TXID.

Maybe we should change a bit a text of help, because of TXID the flag `--structure` flag doesn't seem to be optional.

I've tested `transaction show` command with `--silent` flag and got no output  - truly silent :-)

**Comment by d5000** [comment]
The problem of this command is that I tried to throw two quite different former commands together ;)

With the `--structure` flag, it can be applied to all TXIDs, while without it, it only works with transactions which were stored ("labeled"), above all for the purpose to store a partly signed TX (which can't be stored in the Slimcoin wallet), to use it in the DEX.

Maybe the following change could make sense:

Pacli could search automatically if the given parameter is a label or TXID that was stored in the extended config file, and if yes, then by default it shows the TX HEX of that transaction. If instead a TXID is given which was not stored, then it can be automatically assumed that the user wants to use the `--structure` flag. If the user wants to see the structure of a transaction that was stored/labeled, they could specify `--structure`, but it would no longer be required.

Does this sound better to you?

**Comment by buhtignew** [comment]
> Pacli could search automatically if the given parameter is a label or TXID that was stored in the extended config file, and if yes, then by default it shows the TX HEX of that transaction. 
At the moment it seems like `transaction show LABEL` command outputs the TXID.
Should it output TX HEX or TX HEX is the same thing as TXID?
If it the same thing I was thinking it can output the LABEL or _no label_ message if the `transaction show TXID` command is run and the TXID is the `transaction show LABEL` is run.
Would it make sense?

> If instead a TXID is given which was not stored, then it can be automatically assumed that the user wants to use the --structure flag. 
If the above doesn't make sense this can be a valid alternative.

> If the user wants to see the structure of a transaction that was stored/labeled, they could specify --structure, but it would no longer be required.
I agree here, but I'm not sure to understand why it would be no longer required?

**Comment by d5000** [comment]
> At the moment it seems like transaction show LABEL command outputs the TXID.
> Should it output TX HEX or TX HEX is the same thing as TXID?

It should be output the TX HEX - i.e. the whole transaction in hex format, which is much longer than a TXID. However, there is currently no mechanism that verifies a valid TX HEX has been stored. So actually you can store everything, including a TXID or all kinds of words. I have changed that already, so in the next version it won't be possible anymore to store invalid transactions with "transaction set".

I have now thought a bit how a normal user would probably expect this command to work, and while I was initially reluctant to create a redundancy with the Slimcoin command `getrawtransaction`, I guess the most "usable" scheme would be that if a TXID is given (instead of a LABEL) then the command should show always the transaction without the sender/receiver "structure". So we would have:

`transaction show TXID_OR_LABEL `without anything: If a label/txid was stored in the config file, then show the corresponding TX HEX, otherwise show the TX HEX of the transaction.
`transaction show TXID_OR_LABEL --decode / -d`: Show the transaction JSON instead of the HEX (very similar to `slimcoind decoderawtransaction / getrawtransaction 1`) 
`transaction show TXID --anatomy / -a`: Show the "structure", i.e. senders and receivers.

re: flags:

I think I like `--anatomy` from your proposals, `--participants` is in my view also good but anatomy may be the best candidate.

**Comment by buhtignew** [comment]
> If a label/txid was stored in the config file, then show the corresponding TX HEX, otherwise show the TX HEX of the transaction.

Do you mean that the LABEL is considered as TX HEX by you as well in this context, or you mean that a transaction can be associated with another TX HEX in order to perform which in reality would be a label and that that TX HEX will be used later on in some other operation, i.e. DEX?

**Comment by d5000** [comment]
> Are you aware that at the moment the output is the TXID instead, in case the LABEL is used?

As I'd already written (I think at least ;-) ) it depends what you stored when you used `transaction set` with that label. If you stored a TXID then you will get a TXID when you use `transaction show`. In my case, I get the transaction hex string correctly because I stored the TX HEX string.
As this is not intended, I'll add a check to what you are storing with `transaction set`, so it would no longer accept to store a TXID with transaction set. I'll do that ASAP (probably in the next 2 hours or so, I had forgot it to include it in the update yesterday).
If you still have doubts, please post your output of `transaction list --named`.

> Do you mean that the LABEL is considered as TX HEX by you as well in this context, or you mean that a transaction can be associated with another TX HEX in order to perform which in reality would be a label and that that TX HEX will be used later on in some other operation, i.e. DEX?

I'm not completely understanding where your doubt is, so I try to clarify the terms I use:

- What is stored in the extended config file is a dictionary, with all entries containing a LABEL as the key and a transaction in HEX format (TX HEX) as the value.
- The LABEL is the identifier, like when you store labels for addresses it can be any alphanumeric string. Here is probably the point which can lead to confusion: The label _can_ also be the TXID of the transaction, which in some cases is the intended behaviour - if you want to store a complete, signed transaction but still don't want to post it to the network. If you don't have signed the transaction completely, as is the case in the second step of a DEX exchange, then you should NOT use the TXID as label, because the TXID will change once it is completely signed - an arbitrary string is then better.
- The TX HEX is the transaction in hex format. For obvious reasons, the TX HEX is never used as a label. The TX HEX of a partly signed transaction is used in the DEX, as it will be submitted to the other party to be completely signed. This is the main reason to store transactions in pacli.

> My only objection to anatomy, also in terms of the less astonishment principle, is that the flag -a is usually perceived as "all".

That's true. Maybe we should look for an alternative then, as `--all` is used quite often in pacli also.

**Comment by d5000** [comment]
I reworked both `transaction set `and` transaction show` in a way that only correct transactions can be stored, TXIDs will be marked as invalid. If a transaction was already stored incorrectly, a warning is printed out.

See commit: https://github.com/slimcoin-project/pacli/commit/41ed3b914aba00076620553e20897bee6e44f153

**Comment by buhtignew** [comment]
> > My only objection to anatomy, also in terms of the less astonishment principle, is that the flag -a is usually perceived as "all".
>That's true. Maybe we should look for an alternative then, as --all is used quite often in pacli also.

Maybe `-f` that would mean `--form` or `--formation`? Or `-c` like `--content` or `--composition`? Or `-o` like `--outline`?

**Comment by buhtignew** [comment]
> As I'd already written (I think at least ;-) ) it depends what you stored when you used transaction set with that label. If you stored a TXID then you will get a TXID when you use transaction show. In my case, I get the transaction hex string correctly because I stored the TX HEX string.

I've upgraded pacli to the recent version and now the output I'm receiving is more consistent with what I was expecting basing on your explanations:
1) The `transaction show` output of the transaction I've never labeled seems to be the transaction's hex.
Here is the output of the `transaction show ea2b72f7796a5a294a8fe98be32ffd62575506c42aa22f0501c6f2da25e093e1` command, for example: 
`'010000007eaedb6401000000000000000000000000000000000000000000000000000'
'0000000000000ffffffff0e047eaedb640101062f503253482fffffffff0190226d08'
'00000000232103f0ef2f8702b64da8da1481959a4486b16b94614831e194b2ebfdf8f'
'11c764a85ac00000000'`
2) My output of the `transaction list --named` is:
`{
    'new': 'f4e26f3a14a7fbe7ea45bd60ad9005ebbaa36a3cecd864c7f3ef517c5fe27c86'
}
`
3) The output of `transaction show new` is:
`WARNING: Transaction was not stored correctly.
'f4e26f3a14a7fbe7ea45bd60ad9005ebbaa36a3cecd864c7f3ef517c5fe27c86'`, although it doesn't seems to me to have stored TXID while using `transaction set` command.
4) My output of `transaction show f4e26f3a14a7fbe7ea45bd60ad9005ebbaa36a3cecd864c7f3ef517c5fe27c86` is:
`'01000000e372f36401924ce8783e4b8632ff714add68efac00ea39b70317d255a2d7e'
'd9969c4d87cdc0000000049483045022100f275ae3dedd399657cec59854f22dbc509'
'11ceb164436b49996b9c20b9ac5eb70220629b6cb33fb70b137e40fa626190f2c0586'
'c6f75f5e09825e1b9511f315aad9801ffffffff0610270000000000001976a9143b73'
'a98afce267f672a4c70c6c2ea108b2d0138d88ac10270000000000003b6a390801120'
'f8094ebdc0380cab5ee0180cab5ee01180422220a207d69408f084cbf041a569aafe5'
'0dc7d1a192c24e946d4a889be3ce6edfb4ae9310270000000000001976a914e948818'
'f03eb17d4e862d861b9f21c8f664cf99588ac10270000000000001976a91425c2f39a'
'40ab4a48120c2adb5904584c2ef8d36b88ac10270000000000001976a914b5a2ced9a'
'fd823cb142607e2d4e7e979cb73b87f88acb0e35106000000001976a914388a02daa5'
'287e9d60d84ed901ecaf5d34a563f788ac00000000'` - looks like hex, but it's much longer than that mentioned for another transaction above, maybe because this transaction contains much more info that the other one?
4) The output of `transaction show new --anatomy` is `Error: Transaction does not exist or is corrupted.` - wouldn't it make sense to output the hex as for the same command with TXID instead of the LABEL?

**Comment by buhtignew** [comment]
> I reworked both transaction set and transaction show in a way that only correct transactions can be stored. If TXIDs or other arbitrary strings are stored with this command, they will be marked as invalid (it's not possible to store them with this command anymore). If a transaction was already stored incorrectly, a warning is printed out.

I feel like there is something big I'm not seeing here:
I've run `transaction set second_one 59db10eb75078bc230abc68cbf1e31a9c989c0e6451b1ab78c5b38edc505eb7d` and got `Error: Invalid transaction hex string.`, but I was trying to adhere to the following instruction from help:
`    Usage:

    pacli transaction set LABEL TX

    Stores hex string of transaction TX together with label LABEL.
`
Should I've been using a HEX string or the label is this case? (Because from the command description I'm not quite getting how can I do to store both the hex string and the label, since the hex string is not mentioned in the command I don't know where should I look for it).

**Comment by d5000** [comment]
> Maybe `-f` that would mean `--form` or `--formation`? Or `-c` like `--content`, `--composition`, `--construction`? Or `-o` like `--outline`? Or maybe `-d` like `--design`? Another idea can be `--layout`. Or just `--more`, like I've already suggested earlier.

I think  `--layout` is quite good. I'll consider it.

I've thought also about `--coinflow` or `--moneyflow` but its still not ideal, but could be also a base to brainstorm further.
I've also looked into the Fire documentation if it can be definied that the short mode of the option/flag is another character than the first one (e.g. `--structure` could become `-t`, or `--anatomy` to `-o`) but I've not found that part still.

So maybe I'll go on with `--layout/-l`

> I've upgraded pacli to the recent version and now the output I'm receiving is more consistent with what I was expecting basing on your explanations:

Your output of your points 1 to  4 seem fine. You seem to have stored `new` with the TXID of the transaction, not its content. Try to store the TX HEX entirely (you could, for example, do something like: `pacli transaction set LABEL $(slimcoind getrawtransaction TXID 1)` to simply store an arbitrary transaction you know the TXID of. This is not really the way it's meant to be used but it may show you how the command `transaction show` should work.

The error message in case 5 is actually a "mix" of two different issues:
- First, again you used the label `new` which was connected to a TXID, not to a TX HEX.
- Second, `--anatomy` (or whatever flag will replace it) still doesn't support labels, only TXIDs. I wonder if this is necessary, because the anatomy will be used mostly for transactions which exist on the blockchain, which in my opinion do not really need to be stored with a label with the `transaction set` function. So even if `new` referenced a complete transaction it wouldn't have worked. If you think this function would make sense, then I can add it, it's not a big deal (it could also be beneficial if we take into account the POLA principle).

I'll answer your last questions in a new message :)

**Comment by d5000** [comment]
Now to your last post:

Yes, you got it correct - `transaction set` expects the TXHEX as an argument. I think I should update the docstring to make that clearer.

> Then I edited the four lines of the Hex output into one line
This would have been an example where `--silent` could have benefittet you: it would have printed the hex string without the "beautification" (i.e. colours, line break and quotation marks)

Maybe it would make sense however to offer an alternative mode, which would be exactly what I think you were trying to do first: storing the TX HEX of a transaction existing on the blockchain, giving the TXID as an argument. As TXIDs are quite easy to recognize this could be done without any new flags. I only didn't add that because I didn't think this would have any use cases ... but I can imagine some people wanting to label some arbitrary transactions in some way. for example a burn transaction.

> Reading forth your explanation it seems to me that the average user won't ever have any need to label his transaction. The thing will be done automatically by the DEX because it's mainly the DEX's feature. Have I got it right?

I think you got it right. I still haven't added the "automatic" storage option to the DEX (I'll have to think about a scheme for the labels), but once this is done, this feature would be only of use for power users, for example if you want to set an easy-to-recognize custom label or try different values for DEX transactions.

**Comment by buhtignew** [comment]
> I think --layout is quite good. I'll consider it.

Another one I've thought about it `--blueprint`.

> I've also looked into the Fire documentation if it can be definied that the short mode of the option/flag is another character than the first one (e.g. --structure could become -t, or --anatomy to -o) but I've not found that part still.

I think that keeping the possibility to use the first letter of the word as a flag would help memorization.

**Comment by d5000** [comment]
I think you're right, let's continue the discussion about possible other enhancements of the transaction storage when you're ready to test the DEX.

`--silent` is meant to be used the same way across pacli (in most commands): just print out the raw result (script-friendly) of the command or query without any warnings or additional infos, so I think I'll not change this flag, as it's also quite common in other programs.

**Comment by buhtignew** [comment]
I've got your point.
Just a last suggestion. What if we change it in `--raw` then? (`--silent` doesn't looks too intuitive to me).

**Comment by d5000** [comment]
I've not forgotten that suggestion, and had actually thought about --raw, but in general I like --silent more (it could also be renamed to --quiet/-q, which is used simularly in other CLI applications and doesn't conflict so much with other shortened/1-letter flags).

One of the reasons is that some commands with the --silent flag don't output nothing at all, above all those who store things, where no output is required. The only outputs are the (raw) results of the queries. I wanted to have this flag unified for all commands for which it makes sense. --raw in cases where no output is expected would simply not make sense, and also it would hide the fact that the flag suppresses all warnings and info printouts.

I don't rule out there may be a better name for that flag but I also don't want to spend too much time thinking about it :)

**Comment by buhtignew** [comment]
I like `--quiet` if you think it's appropriate I think it would be nice to have it implemented.

**Comment by d5000** [comment]
Good, I'll go for --quiet then, but that can take more than a day with all commands, so probably if there are more bugs to be fixed in the next days (just fixed the checkpoint issue) this change will occur step by step.

**Comment by buhtignew** [comment]
Yes this issue is solved.
